### ISSUE TITLE
Offline mode

### ISSUE THREAD

**Issue by mspl13** [albalitz/save-android#11]
When the API-mode is enabled and the device is in offline mode/can not connect to the server (for whatever reason) the app crashes. This problem could for example be solved by having a system that queues the changes to the list and synchronizes them when a connection is available again.

**Steps to reproduce**:
1. Enable API-mode.
2. Disable network connection (e.g. by putting the device into airplane mode).
3. Share a link to the Save app via the android share menu.
4. Wait and see the app crash.

**Comment by albalitz** [comment]
Thanks for reporting this.
I'll work on this when I find the time. ;)

**Comment by albalitz** [comment]
@mspl, you suggested
> a system that queues the changes to the list and synchronizes them when a connection is available again

What do you think would be a nicer approach from a usability standpoint?
1. queueing saved links and retrying automatically until it works, or
1. queueing saved links and showing a notification with a 'try again' button

**Comment by mspl13** [comment]
What about a third approach: queuing saved link and retrying automatically when the connectivity status changes to `connected`.

I think the [ConnectivityManager](https://developer.android.com/training/monitoring-device-state/connectivity-monitoring.html) would be the right thing to use here.

**Comment by albalitz** [comment]
Something like that is what I meant with the first approach. :)
I'll take a look at the ConnectivityManager later - thanks for the link.

_Sent from my Motorola XT1092 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by albalitz** [comment]
| Trying to save without connection | Connection is back |
| ------------------------------------------ | --- |
| ![screenshot_20170816-172512](https://user-images.githubusercontent.com/9308749/29371477-ba60cb00-82a8-11e7-8346-2ac121601b1a.png) | ![screenshot_20170816-172523](https://user-images.githubusercontent.com/9308749/29371489-be6d9d7c-82a8-11e7-9fa9-a3b2a91dd3bc.png) |

This should work when the connection comes back soon after queueing. Haven't tested if returning connection is detected correctly when the app isn't running for a while.
Alternatively, when opening the app, queued links are sent to the API.

**Comment by albalitz** [comment]
Something's not working yet.
Notes to self:
- does queueing and then saving multiple links work?
- does dropping a link work?

**Comment by albalitz** [comment]
So, @mspl13, I decided to change the queue to not save automatically, as I found that too unstable. Instead, I added a fancy button that's displayed over the list, when there are queued links (and a network connection is available).

Here's a screenshot with the button:
<img width="50%" src="https://user-images.githubusercontent.com/9308749/31299183-83272f5e-aaed-11e7-9cd7-91d74defd82c.png" />

On my phone this works better now. I'll close this issue again for now. If there's still issues with this, let me know. :)
